File: android/aspectly-bridge/src/test/kotlin/com/aspectly/bridge/BridgeHostTest.kt
package com.aspectly.bridge

import com.aspectly.bridge.protocol.BridgeErrorType
import com.google.gson.JsonElement
import com.google.gson.JsonParser
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class BridgeHostTest {
    private lateinit var mock: MockBrowserBridge
    private lateinit var bridge: BridgeHost

    @Before
    fun setUp() {
        mock = MockBrowserBridge()
        bridge = BridgeHost(mock)
    }

    @After
    fun tearDown() {
        bridge.dispose()
    }

    // region Handler registration

    @Test
    fun registerHandlerAddsHandler() {
        bridge.registerHandler("test") { "result" }
        assertTrue(bridge.registeredMethods.contains("test"))
    }

    @Test
    fun unregisterHandlerRemovesHandler() {
        bridge.registerHandler("test") { "result" }
        bridge.unregisterHandler("test")
        assertFalse(bridge.registeredMethods.contains("test"))
    }

    @Test
    fun registerHandlerWithSameNameOverwrites() = runBlocking {
        bridge.registerHandler("test") { "first" }
        bridge.registerHandler("test") { "second" }

        bridge.processMessage(request("test", "1"))

        assertTrue(mock.lastScript()!!.contains("second"))
    }

    // endregion

    // region Init handshake

    @Test
    fun processMessageWithInitUpdatesSupportedMethods() = runBlocking {
        bridge.processMessage(initMessage(listOf("jsMethod1", "jsMethod2")))

        assertTrue(bridge.supportedMethods.contains("jsMethod1"))
        assertTrue(bridge.supportedMethods.contains("jsMethod2"))
    }

    @Test
    fun handleInitOnlySendsInitResultNotOurInit() = runBlocking {
        bridge.registerHandler("myMethod") { "result" }

        bridge.processMessage(initMessage(listOf("jsMethod")))

        assertEquals(1, mock.sentScripts.size)
        assertTrue(mock.sentScripts[0].contains("InitResult"))
        assertFalse(mock.sentScripts[0].contains("""type\":\"Init\""""))
    }

    @Test
    fun initializesOnlyWhenBothInitAndInitResultReceived() = runBlocking {
        var eventFired = false
        bridge.onInitialized = { eventFired = true }

        bridge.processMessage(bridgeMessage("InitResult", "true"))
        assertFalse(bridge.isInitialized)
        assertFalse(eventFired)

        bridge.processMessage(initMessage(listOf("jsMethod")))
        assertTrue(bridge.isInitialized)
        assertTrue(eventFired)
    }

    @Test
    fun initializeWaitsForBothInitAndInitResult() = runBlocking {
        bridge.handshake(mock, listOf("jsMethod"))
        assertTrue(bridge.isInitialized)
        assertTrue(bridge.supportedMethods.contains("jsMethod"))
    }

    @Test
    fun initializeWithHandlersRegistersAndInits() = runBlocking {
        // Handlers must be typed as RawBridgeHandler so the lambda is inferred as
        // a `suspend` function (Kotlin does not coerce bare lambdas to suspend
        // inside a Pair/map literal).
        val h1: RawBridgeHandler = { _ -> "result1" }
        val h2: RawBridgeHandler = { _ -> "result2" }
        coroutineScope {
            val job = async {
                bridge.initialize(mapOf("method1" to h1, "method2" to h2))
            }
            waitFor { bridge.registeredMethods.size >= 2 }
            assertTrue(bridge.registeredMethods.contains("method1"))
            assertTrue(bridge.registeredMethods.contains("method2"))

            bridge.processMessage(initMessage(listOf("jsMethod")))
            bridge.processMessage(bridgeMessage("InitResult", "true"))
            job.await()
        }
        assertTrue(bridge.isInitialized)
    }

    // endregion

    // region Incoming requests

    @Test
    fun processMessageWithRequestCallsHandler() = runBlocking {
        var called = false
        bridge.registerHandler("testMethod") {
            called = true
            mapOf("success" to true)
        }

        bridge.processMessage(request("testMethod", "1"))
        assertTrue(called)
    }

    @Test
    fun processMessageWithUnknownMethodSendsError() = runBlocking {
        bridge.processMessage(request("unknownMethod", "1"))
        assertTrue(mock.lastScript()!!.contains("UNSUPPORTED_METHOD"))
    }

    @Test
    fun requestHandlerReturnsDataSendsSuccessResult() = runBlocking {
        bridge.registerHandler("testMethod") { mapOf("value" to 42) }

        bridge.processMessage(request("testMethod", "test-123"))

        val script = mock.lastScript()!!
        assertTrue(script.contains("Success"))
        assertTrue(script.contains("42"))
    }

    @Test
    fun requestHandlerThrowsSendsRejectedError() = runBlocking {
        bridge.registerHandler("testMethod") { throw RuntimeException("test error") }

        bridge.processMessage(request("testMethod", "test-123"))

        val script = mock.lastScript()!!
        assertTrue(script.contains("REJECTED"))
        assertTrue(script.contains("test error"))
    }

    @Test
    fun handleRequestTimesOutSlowHandler() = runBlocking {
        val shortBridge = BridgeHost(mock, timeoutMs = 100)
        shortBridge.registerHandler("slowMethod") {
            delay(5000)
            mapOf("value" to "late")
        }

        shortBridge.processMessage(request("slowMethod", "1"))

        assertTrue(mock.lastScript()!!.contains("METHOD_EXECUTION_TIMEOUT"))
        shortBridge.dispose()
    }

    @Test
    fun typedHandlerDecodesParams() = runBlocking {
        bridge.registerTypedHandler<AddParams>("add") { p -> p.a + p.b }

        bridge.processMessage(
            bridgeMessage("Request", """{"method":"add","request_id":"1","params":{"a":10,"b":20}}""")
        )

        val wrapper = parseSentScript(mock.lastScript()!!)
        val data = wrapper.asJsonObject.getAsJsonObject("event").getAsJsonObject("data").get("data")
        assertEquals(30, data.asInt)
    }

    // endregion

    // region Sending requests

    @Test
    fun sendWithSuccessResultResolvesPendingRequest() = runBlocking {
        bridge.handshake(mock, listOf("jsMethod"))
        coroutineScope {
            val job = async { bridge.send<String>("jsMethod") }
            waitFor { mock.anyScriptContains("jsMethod") }
            bridge.processMessage(successResult("\"test result\"", "1"))
            assertEquals("test result", job.await())
        }
    }

    @Test
    fun sendWithErrorResultRejectsWithBridgeException() = runBlocking {
        bridge.handshake(mock, listOf("jsMethod"))
        coroutineScope {
            val job = async { runCatching { bridge.send<String>("jsMethod") } }
            waitFor { mock.anyScriptContains("jsMethod") }
            bridge.processMessage(
                bridgeMessage(
                    "Result",
                    """{"type":"Error","error":{"error_type":"REJECTED","error_message":"Method failed"},"request_id":"1"}"""
                )
            )
            val result = job.await()
            assertTrue(result.isFailure)
            val ex = result.exceptionOrNull()
            assertTrue(ex is BridgeException)
            assertEquals(BridgeErrorType.REJECTED, (ex as BridgeException).errorType)
            assertEquals("Method failed", ex.message)
        }
    }

    @Test
    fun sendWithTimeoutThrowsTimeoutError() = runBlocking {
        bridge.handshake(mock, listOf("jsMethod"))
        val result = runCatching { bridge.send<String>("jsMethod", timeoutMs = 100) }
        assertTrue(result.isFailure)
        assertEquals(
            BridgeErrorType.METHOD_EXECUTION_TIMEOUT,
            (result.exceptionOrNull() as BridgeException).errorType
        )
    }

    @Test
    fun sendBeforeInitializationThrows() = runBlocking {
        val result = runCatching { bridge.send<String>("anyMethod") }
        assertTrue(result.isFailure)
        assertEquals(
            BridgeErrorType.BRIDGE_NOT_AVAILABLE,
            (result.exceptionOrNull() as BridgeException).errorType
        )
    }

    @Test
    fun sendUnsupportedMethodThrows() = runBlocking {
        bridge.handshake(mock, listOf("otherMethod"))
        val result = runCatching { bridge.send<String>("unknownMethod") }
        assertTrue(result.isFailure)
        assertEquals(
            BridgeErrorType.UNSUPPORTED_METHOD,
            (result.exceptionOrNull() as BridgeException).errorType
        )
    }

    // endregion

    // region Ignored messages

    @Test
    fun processMessageWithEmptyStringIsIgnored() = runBlocking {
        bridge.processMessage("")
        assertTrue(mock.sentScripts.isEmpty())
    }

    @Test
    fun processMessageWithInvalidJsonIsIgnored() = runBlocking {
        bridge.processMessage("not json {")
        assertTrue(mock.sentScripts.isEmpty())
    }

    @Test
    fun processMessageWithNonBridgeEventIsIgnored() = runBlocking {
        bridge.processMessage("""{"type":"OtherEvent","data":{}}""")
        assertTrue(mock.sentScripts.isEmpty())
    }

    // endregion

    // region Dispose

    @Test
    fun disposeDisposesBrowserBridge() {
        bridge.dispose()
        assertTrue(mock.disposed)
    }

    @Test
    fun disposeCalledTwiceDoesNotThrow() {
        bridge.dispose()
        bridge.dispose()
    }

    @Test
    fun disposeCancelsPendingRequests() = runBlocking {
        bridge.handshake(mock, listOf("jsMethod"))
        coroutineScope {
            val job = async { runCatching { bridge.send<String>("jsMethod") } }
            waitFor { mock.anyScriptContains("jsMethod") }
            bridge.dispose()
            val result = job.await()
            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull() is CancellationException)
        }
    }

    // endregion

    // region Helpers

    data class AddParams(val a: Int, val b: Int)

    private suspend fun BridgeHost.handshake(mock: MockBrowserBridge, jsMethods: List<String>) = coroutineScope {
        val job = async { initialize() }
        waitFor { mock.sentScripts.isNotEmpty() }
        processMessage(initMessage(jsMethods))
        processMessage(bridgeMessage("InitResult", "true"))
        job.await()
    }

    private suspend fun waitFor(timeoutMs: Long = 3000, condition: () -> Boolean) {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (!condition() && System.currentTimeMillis() < deadline) {
            delay(2)
        }
    }

    private fun bridgeMessage(type: String, dataJson: String): String =
        """{"type":"BridgeEvent","event":{"type":"$type","data":$dataJson}}"""

    private fun initMessage(methods: List<String>): String {
        val arr = methods.joinToString(",") { "\"$it\"" }
        return bridgeMessage("Init", """{"methods":[$arr]}""")
    }

    private fun request(method: String, id: String): String =
        bridgeMessage("Request", """{"method":"$method","request_id":"$id","params":{}}""")

    private fun successResult(dataJson: String, id: String): String =
        bridgeMessage("Result", """{"type":"Success","data":$dataJson,"request_id":"$id"}""")

    private fun parseSentScript(script: String): JsonElement {
        val start = script.indexOf("window.postMessage(") + "window.postMessage(".length
        val end = script.indexOf(", '*')", start)
        val literal = script.substring(start, end)
        // literal is a JSON string token; decode it to get the inner JSON.
        val innerJson = JsonParser.parseString(literal).asString
        return JsonParser.parseString(innerJson)
    }

    // endregion
}

File: dart/test/bridge_host_test.dart
import 'dart:convert';

import 'package:aspectly_bridge/aspectly_bridge.dart';
import 'package:test/test.dart';

class MockBrowserBridge implements BrowserBridge {
  @override
  bool get isReady => true;

  void Function(String message)? _onMessage;
  @override
  set onMessage(void Function(String message)? cb) => _onMessage = cb;
  @override
  void Function(String message)? get onMessage => _onMessage;

  final List<String> sentScripts = <String>[];
  bool disposed = false;

  @override
  Future<void> executeScript(String script) async {
    sentScripts.add(script);
  }

  @override
  void dispose() {
    disposed = true;
  }

  bool anyScriptContains(String s) => sentScripts.any((e) => e.contains(s));
}

String bridgeMessage(String type, dynamic data) => jsonEncode(<String, dynamic>{
      'type': 'BridgeEvent',
      'event': <String, dynamic>{'type': type, 'data': data},
    });

Future<void> handshake(
    BridgeHost bridge, MockBrowserBridge mock, List<String> jsMethods) async {
  final initFuture = bridge.initialize();
  await Future<void>.delayed(Duration.zero);
  await bridge.processMessage(bridgeMessage('Init', <String, dynamic>{'methods': jsMethods}));
  await bridge.processMessage(bridgeMessage('InitResult', true));
  await initFuture;
}

void main() {
  late MockBrowserBridge mock;
  late BridgeHost bridge;

  setUp(() {
    mock = MockBrowserBridge();
    bridge = BridgeHost(mock);
  });

  tearDown(() => bridge.dispose());

  group('handler registration', () {
    test('registerHandler adds handler', () {
      bridge.registerHandler('test', (p) => 'result');
      expect(bridge.registeredMethods, contains('test'));
    });

    test('unregisterHandler removes handler', () {
      bridge.registerHandler('test', (p) => 'result');
      bridge.unregisterHandler('test');
      expect(bridge.registeredMethods, isNot(contains('test')));
    });

    test('register with same name overwrites', () async {
      bridge.registerHandler('test', (p) => 'first');
      bridge.registerHandler('test', (p) => 'second');
      await bridge.processMessage(bridgeMessage('Request',
          <String, dynamic>{'method': 'test', 'request_id': '1', 'params': <String, dynamic>{}}));
      expect(mock.sentScripts.last, contains('second'));
    });
  });

  group('init handshake', () {
    test('Init updates supported methods', () async {
      await bridge.processMessage(
          bridgeMessage('Init', <String, dynamic>{'methods': <String>['a', 'b']}));
      expect(bridge.supportedMethods, containsAll(<String>['a', 'b']));
    });

    test('handleInit only sends InitResult, not our Init', () async {
      bridge.registerHandler('myMethod', (p) => 'r');
      await bridge.processMessage(
          bridgeMessage('Init', <String, dynamic>{'methods': <String>['jsMethod']}));
      expect(mock.sentScripts.length, 1);
      expect(mock.sentScripts[0], contains('InitResult'));
    });

    test('initializes only when both Init and InitResult received', () async {
      var fired = false;
      bridge.onInitialized = () => fired = true;
      await bridge.processMessage(bridgeMessage('InitResult', true));
      expect(bridge.isInitialized, isFalse);
      expect(fired, isFalse);
      await bridge.processMessage(
          bridgeMessage('Init', <String, dynamic>{'methods': <String>['jsMethod']}));
      expect(bridge.isInitialized, isTrue);
      expect(fired, isTrue);
    });

    test('initialize waits for both Init and InitResult', () async {
      await handshake(bridge, mock, <String>['jsMethod']);
      expect(bridge.isInitialized, isTrue);
      expect(bridge.supportedMethods, contains('jsMethod'));
    });

    test('initialize(handlers) registers and inits', () async {
      final initFuture = bridge.initialize(<String, BridgeHandler>{
        'm1': (p) => 'r1',
        'm2': (p) => 'r2',
      });
      await Future<void>.delayed(Duration.zero);
      expect(bridge.registeredMethods, containsAll(<String>['m1', 'm2']));
      await bridge.processMessage(
          bridgeMessage('Init', <String, dynamic>{'methods': <String>['jsMethod']}));
      await bridge.processMessage(bridgeMessage('InitResult', true));
      await initFuture;
      expect(bridge.isInitialized, isTrue);
    });
  });

  group('incoming requests', () {
    test('calls handler', () async {
      var called = false;
      bridge.registerHandler('m', (p) {
        called = true;
        return <String, dynamic>{'ok': true};
      });
      await bridge.processMessage(bridgeMessage('Request',
          <String, dynamic>{'method': 'm', 'request_id': '1', 'params': <String, dynamic>{}}));
      expect(called, isTrue);
    });

    test('unknown method sends UNSUPPORTED_METHOD', () async {
      await bridge.processMessage(bridgeMessage('Request',
          <String, dynamic>{'method': 'nope', 'request_id': '1', 'params': <String, dynamic>{}}));
      expect(mock.sentScripts.last, contains('UNSUPPORTED_METHOD'));
    });

    test('handler returns data -> Success', () async {
      bridge.registerHandler('m', (p) => <String, dynamic>{'value': 42});
      await bridge.processMessage(bridgeMessage('Request',
          <String, dynamic>{'method': 'm', 'request_id': '1', 'params': <String, dynamic>{}}));
      expect(mock.sentScripts.last, contains('Success'));
      expect(mock.sentScripts.last, contains('42'));
    });

    test('handler throws -> REJECTED', () async {
      bridge.registerHandler('m', (p) => throw Exception('boom'));
      await bridge.processMessage(bridgeMessage('Request',
          <String, dynamic>{'method': 'm', 'request_id': '1', 'params': <String, dynamic>{}}));
      expect(mock.sentScripts.last, contains('REJECTED'));
      expect(mock.sentScripts.last, contains('boom'));
    });

    test('typed params reach handler', () async {
      bridge.registerHandler('add', (p) => p['a'] + p['b']);
      await bridge.processMessage(bridgeMessage('Request', <String, dynamic>{
        'method': 'add',
        'request_id': '1',
        'params': <String, dynamic>{'a': 10, 'b': 20},
      }));
      expect(mock.sentScripts.last, contains('30'));
    });
  });

  group('sending requests', () {
    test('success resolves', () async {
      await handshake(bridge, mock, <String>['jsMethod']);
      final f = bridge.send<String>('jsMethod');
      await Future<void>.delayed(Duration.zero);
      await bridge.processMessage(bridgeMessage('Result', <String, dynamic>{
        'type': 'Success',
        'data': 'test result',
        'request_id': '1',
      }));
      expect(await f, 'test result');
    });

    test('error rejects with BridgeException', () async {
      await handshake(bridge, mock, <String>['jsMethod']);
      final f = bridge.send<String>('jsMethod');
      final expectation = expectLater(
        f,
        throwsA(isA<BridgeException>()
            .having((e) => e.errorType, 'errorType', BridgeErrorType.rejected)
            .having((e) => e.message, 'message', 'failed')),
      );
      await Future<void>.delayed(Duration.zero);
      await bridge.processMessage(bridgeMessage('Result', <String, dynamic>{
        'type': 'Error',
        'error': <String, dynamic>{'error_type': 'REJECTED', 'error_message': 'failed'},
        'request_id': '1',
      }));
      await expectation;
    });

    test('timeout throws', () async {
      await handshake(bridge, mock, <String>['jsMethod']);
      expect(
        () => bridge.send<String>('jsMethod', null, 50),
        throwsA(isA<BridgeException>()
            .having((e) => e.errorType, 'errorType', BridgeErrorType.methodExecutionTimeout)),
      );
    });

    test('send before init throws', () async {
      expect(
        () => bridge.send<String>('m'),
        throwsA(isA<BridgeException>()
            .having((e) => e.errorType, 'errorType', BridgeErrorType.bridgeNotAvailable)),
      );
    });

    test('unsupported method throws', () async {
      await handshake(bridge, mock, <String>['other']);
      expect(
        () => bridge.send<String>('m'),
        throwsA(isA<BridgeException>()
            .having((e) => e.errorType, 'errorType', BridgeErrorType.unsupportedMethod)),
      );
    });
  });

  group('ignored messages', () {
    test('empty', () async {
      await bridge.processMessage('');
      expect(mock.sentScripts, isEmpty);
    });
    test('invalid json', () async {
      await bridge.processMessage('not json {');
      expect(mock.sentScripts, isEmpty);
    });
    test('non-bridge event', () async {
      await bridge.processMessage('{"type":"Other","data":{}}');
      expect(mock.sentScripts, isEmpty);
    });
  });

  group('dispose', () {
    test('disposes browser bridge', () {
      bridge.dispose();
      expect(mock.disposed, isTrue);
    });
    test('called twice does not throw', () {
      bridge.dispose();
      bridge.dispose();
    });
  });
}
